kstatus/wait/wait.go

```diff
@@ -0,0 +1,222 @@
+	statusComputeFunc func(u *unstructured.Unstructured) (*status.Result, error)
+}
+
+// NewResolver creates a new resolver with the provided client. Fetching
+// and polling of resources will be done using the provided client.
+func NewResolver(client client.Reader) *Resolver {
+	return &Resolver{
+		client:            client,
+		statusComputeFunc: status.Compute,
+	}
+}
+
+// ResourceResult is the status result for a given resource. It provides
+// information about the resource if the request was successful and an
+// error if something went wrong.
+type ResourceResult struct {
+	Result *status.Result
+
+	Resource ResourceIdentifier
+
+	Error error
+}
+
+// FetchAndResolve returns the status for a list of resources. It will return
+// the status for each of them individually. The slice of ResourceIdentifiers will
+// only be used to get the information needed to fetch the updated state of
+// the resources from the cluster.
+func (r *Resolver) FetchAndResolve(ctx context.Context, resources []ResourceIdentifier) []ResourceResult {
+	var results []ResourceResult
+
+	for _, resource := range resources {
+		u, err := r.fetchResource(ctx, resource)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				results = append(results, ResourceResult{
+					Resource: resource,
+					Result: &status.Result{
+						Status:  status.CurrentStatus,
+						Message: "Resource does not exist",
+					},
+				})
+			} else {
+				results = append(results, ResourceResult{
+					Result: &status.Result{
+						Status:  status.UnknownStatus,
+						Message: fmt.Sprintf("Error fetching resource from cluster: %v", err),
+					},
+					Resource: resource,
+					Error:    err,
+				})
+			}
+			continue
+		}
+		res, err := r.statusComputeFunc(u)
+		results = append(results, ResourceResult{
+			Result:   res,
+			Resource: resource,
+			Error:    err,
+		})
+	}
+
+	return results
+}
+
+// Event is returned through the channel returned after a call
+// to WaitForStatus. It contains an update to either an individual
+// resource or to the aggregate status for the set of resources.
+type Event struct {
+	// Type defines which type of event this is.
+	Type EventType
+
+	// AggregateStatus is the aggregated status for all the provided resources.
+	AggregateStatus status.Status
+
+	// EventResource is information about the event to which this event pertains.
+	// This is only populated for ResourceUpdate events.
+	EventResource *EventResource
+}
+
+type EventType string
+
+const (
+	// The status/message for a resource has changed. This also means the
+	// aggregate status might have changed.
+	ResourceUpdate EventType = "ResourceUpdate"
+
+	// All resources has reached the current status.
+	Completed EventType = "Completed"
+
+	// The wait was stopped before all resources could reach the
+	// Current status.
+	Aborted EventType = "Aborted"
+)
+
+// EventResource contains information about the resource for which
+// a specific Event pertains.
+type EventResource struct {
```

Comment:
ResourceStatusEvent?

Reply:
I wanted to avoid using the word status in most of the types since it seems redundant as everything is related to status. Also, this type isn't an event in itself, it contains information about which resource an event is associated to. So I think I prefer the current name here.